theseus/embodied/motionmodel/misc.py

```diff
@@ -0,0 +1,160 @@
+            if self.dims is None
+            else base_error[:, self.dims].view(-1, self.dim())
+        )
+        return error, below_idx, above_idx
+
+    def error(self) -> torch.Tensor:
+        return self._compute_error()[0]
+
+    def jacobians(self) -> Tuple[List[torch.Tensor], torch.Tensor]:
+        error, below_idx, above_idx = self._compute_error()
+        raw_J = self.vector.tensor.new_zeros(
+            self.vector.shape[0], self.vector.dof(), self.vector.dof()
+        )
+        if self.side in ["below", "both"]:
+            raw_J[below_idx.diag_embed()] = -1.0
+        if self.side in ["above", "both"]:
+            raw_J[above_idx.diag_embed()] = 1.0
+
+        J = (
+            raw_J
+            if self.dims is None
+            else raw_J[:, self.dims, :].view(-1, self.dim(), self.vector.dof())
+        )
+
+        return [J], error
+
+    def _copy_impl(self, new_name: Optional[str] = None) -> "HingeLoss":
+        return HingeLoss(
+            self.vector.copy(),
+            self.limit.copy(),
+            self.threshold.copy(),
+            self.weight.copy(),
+            name=new_name,
+            side=self.side,
+            dims=self.dims,
+        )
+
+
+class Nonholonomic(CostFunction):
+    def __init__(
+        self,
+        pose: Union[SE2, Point3, Vector],
+        vel: Union[Point3, Vector],
+        cost_weight: CostWeight,
+        name: Optional[str] = None,
+    ):
+        super().__init__(cost_weight, name=name)
+        if vel.dof() != 3 or pose.dof() != 3:
+            raise ValueError(
+                "Nonholonomic only accepts 3D velocity or poses (x, y, z dims). "
+                "Poses can either be SE2 or Vector variables. Velocities only Vector."
+            )
+        self.pose = pose
+        self.vel = vel
+        self.register_optim_vars(["pose", "vel"])
+        self.weight = cost_weight
+
+    def dim(self):
+        return 1
+
+    def _compute_error(
+        self,
+    ) -> Tuple[torch.Tensor, Optional[torch.Tensor], Optional[torch.Tensor]]:
+        vel = self.vel.tensor
+        if isinstance(self.pose, SE2):
+            error = vel[:, 1]
+            cos, sin = None, None
+        else:
+            cos = self.pose[:, 2].cos()
+            sin = self.pose[:, 2].sin()
+            error = vel[:, 1] * cos - vel[:, 0] * sin
+        return error.view(-1, 1), cos, sin
+
+    def error(self) -> torch.Tensor:
+        return self._compute_error()[0]
+
+    def jacobians(self) -> Tuple[List[torch.Tensor], torch.Tensor]:
+        # Pre-allocate jacobian tensors
+        batch_size = self.vel.shape[0]
+        Jvel = self.vel.tensor.new_zeros(batch_size, 1, 3)
+        error, cos, sin = self._compute_error()
+        if isinstance(self.pose, SE2):
+            Jpose = error.new_zeros(error.shape[0], 1, 3)
+            Jvel = error.new_zeros(error.shape[0], 1, 3)
+            Jvel[:, 0, 1] = 1
+        else:
+            Jpose = self.vel.tensor.new_zeros(self.vel.shape[0], 1, 3)
```

Comment:
Does this if/else branching make it incompatible with our vectorization?

Reply:
Not sure to be honest, but I could rewrite this in a way that this branching happens at construction time, maybe safer.

Reply:
Fixed in 9a7cb885bfffa9bdb2fd8d41c77bf6563f61e666.